src/app/clusters/mode-base-server/mode-base-server.cpp

```diff
@@ -0,0 +1,445 @@
+    {
+        // Read the StartUpMode attribute and set the CurrentMode attribute
+        // The StartUpMode attribute SHALL define the desired startup behavior of a
+        // device when it is supplied with power and this state SHALL be
+        // reflected in the CurrentMode attribute.  The values of the StartUpMode
+        // attribute are listed below.
+
+        DataModel::Nullable<uint8_t> startUpMode = GetStartUpMode();
+        if (!startUpMode.IsNull())
+        {
+#ifdef EMBER_AF_PLUGIN_ON_OFF
+            // OnMode with Power Up
+            // If the On/Off feature is supported and the On/Off cluster attribute StartUpOnOff is present, with a
+            // value of On (turn on at power up), then the CurrentMode attribute SHALL be set to the OnMode attribute
+            // value when the server is supplied with power, except if the OnMode attribute is null.
+            if (emberAfContainsServer(mEndpointId, OnOff::Id) &&
+                emberAfContainsAttribute(mEndpointId, OnOff::Id, OnOff::Attributes::StartUpOnOff::Id) &&
+                emberAfContainsAttribute(mEndpointId, mClusterId, ModeBase::Attributes::OnMode::Id) &&
+                HasFeature(ModeBase::Feature::kOnOff))
+            {
+                DataModel::Nullable<uint8_t> onMode = GetOnMode();
+                bool onOffValueForStartUp           = false;
+                if (!emberAfIsKnownVolatileAttribute(mEndpointId, OnOff::Id, OnOff::Attributes::StartUpOnOff::Id) &&
+                    OnOffServer::Instance().getOnOffValueForStartUp(mEndpointId, onOffValueForStartUp) == EMBER_ZCL_STATUS_SUCCESS)
+                {
+                    if (onOffValueForStartUp && !onMode.IsNull())
+                    {
+                        ChipLogDetail(Zcl, "ModeBase: CurrentMode is overwritten by OnMode");
+                        // it is overwritten by the on/off cluster.
+                        return CHIP_NO_ERROR;
+                    }
+                }
+            }
+#endif // EMBER_AF_PLUGIN_ON_OFF
+
+            BootReasonType bootReason = BootReasonType::kUnspecified;
+            CHIP_ERROR error          = DeviceLayer::GetDiagnosticDataProvider().GetBootReason(bootReason);
+
+            if (error != CHIP_NO_ERROR)
+            {
+                ChipLogError(Zcl, "Unable to retrieve boot reason: %" CHIP_ERROR_FORMAT, error.Format());
+                // We really only care whether the boot reason is OTA.  Assume it's not.
+                bootReason = BootReasonType::kUnspecified;
+            }
+            if (bootReason == BootReasonType::kSoftwareUpdateCompleted)
+            {
+                ChipLogDetail(Zcl, "ModeBase: CurrentMode is ignored for OTA reboot");
+                return CHIP_NO_ERROR;
+            }
+
+            // Initialise currentMode to 0
+            uint8_t currentMode = GetCurrentMode();
+
+            if (startUpMode.Value() != currentMode)
```

Comment:
The code here has changed since the implementation of persistent storage. It now exists in the `loadPersistentAttributes` function. Do you mean that when we read the persistent value at init, we should also call the `MatterReportingAttributeChangeCallback`? Which we can do by using the `UpdateCurrentMode` function.

Reply:
I assume that is what you meant, because it makes sense. It is changed [here](https://github.com/project-chip/connectedhomeip/pull/27504/commits/a83ee26268ccb436d914fc9ff5e89e5e0e96f558).